eth/eth2.go

```diff
@@ -0,0 +1,325 @@
+
+		// Execute the transaction
+		api.env.state.Prepare(tx.Hash(), common.Hash{}, api.env.tcount)
+		err := api.commitTransaction(tx, coinbase)
+		switch err {
+		case core.ErrGasLimitReached:
+			// Pop the current out-of-gas transaction without shifting in the next from the account
+			log.Trace("Gas limit exceeded for current block", "sender", from)
+			txs.Pop()
+
+		case core.ErrNonceTooLow:
+			// New head notification data race between the transaction pool and miner, shift
+			log.Trace("Skipping transaction with low nonce", "sender", from, "nonce", tx.Nonce())
+			txs.Shift()
+
+		case core.ErrNonceTooHigh:
+			// Reorg notification data race between the transaction pool and miner, skip account =
+			log.Trace("Skipping account with high nonce", "sender", from, "nonce", tx.Nonce())
+			txs.Pop()
+
+		case nil:
+			// Everything ok, collect the logs and shift in the next transaction from the same account
+			api.env.tcount++
+			txs.Shift()
+			transactions = append(transactions, tx)
+
+		default:
+			// Strange error, discard the transaction and get the next in line (note, the
+			// nonce-too-high clause will prevent us from executing in vain).
+			log.Debug("Transaction failed, account skipped", "hash", tx.Hash(), "err", err)
+			txs.Shift()
+		}
+	}
+
+	block, err := api.eth.Engine().FinalizeAndAssemble(bc, header, api.env.state, transactions, nil /* uncles */, api.env.receipts)
+	if err != nil {
+		return nil, err
+	}
+
+	var logs []*types.Log
+	var receipts = make(types.Receipts, len(api.env.receipts))
+	hash := block.Hash()
+	for i, receipt := range api.env.receipts {
+		// add block location fields
+		receipt.BlockHash = hash
+		receipt.BlockNumber = block.Number()
+		receipt.TransactionIndex = uint(i)
+
+		receipts[i] = new(types.Receipt)
+		*receipts[i] = *receipt
+		// Update the block hash in all logs since it is now available and not when the
+		// receipt/log of individual transactions were created.
+		for _, log := range receipt.Logs {
+			log.BlockHash = hash
+		}
+		logs = append(logs, receipt.Logs...)
+	}
+
+	block.Header().ReceiptHash = types.DeriveSha(receipts, new(trie.Trie))
+
+	return &ExecutableData{
+		Coinbase:     block.Coinbase(),
+		StateRoot:    block.Root(),
+		GasLimit:     block.GasLimit(),
+		GasUsed:      block.GasUsed(),
+		Transactions: []*types.Transaction(block.Transactions()),
+		ReceiptRoot:  block.ReceiptHash(),
+		LogsBloom:    block.Bloom().Bytes(),
+		BlockHash:    block.Hash(),
+		Difficulty:   block.Difficulty(),
+		ParentHash:   block.ParentHash(),
+	}, nil
+}
+
+// Structure described at https://hackmd.io/T9x2mMA4S7us8tJwEB3FDQ
+type NewBlockParams struct {
+	RandaoMix              common.Hash    `json:"randao_mix"`
+	Slot                   uint64         `json:"slot"`
+	Timestamp              uint64         `json:"timestamp"`
+	RecentBeaconBlockRoots []common.Hash  `json:"recent_beacon_block_roots"`
+	ExecutableData         ExecutableData `json:"executable_data"`
+}
```

Comment:
`RandaoMix`, `Slot` and `RecentBeaconBlockRoots` seems unused. I guess they are removed from the spec?

Reply:
These are useful to have when implementing the opcodes (which were present in previous catalyst versions). It was indeed removed from the latest spec for simplicity of the first devnet.

You can find the latest spec for RPC here: https://notes.ethereum.org/@n0ble/rayonism-the-merge-spec

Reply:
Right, according to the spec there is also `ExecutionPayload` object in the params of this method, https://notes.ethereum.org/@n0ble/rayonism-the-merge-spec#Parameters1